### PR TITLE
Erg #16 refactoring drag drop algorithm

### PR DESCRIPTION
**Pull request by bleto**
- acceleration of the algorithm, 
- improvement of the relationship between categories,

### CODE REVIEW

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 1]
```diff
@@ -45,202 +45,142 @@ export default {
...
         dataWithoutGhostElement() {
-            return this.gridData.filter(element => element.id !== this.ghostIndex);
-        },
-        ghostElement() {
-            // Last row might be 0 - that's why === null
-            return this.lastRow === null ? null : this.gridData[this.lastRow];
+            return this.gridData.filter(element => element.id !== this.ghostElement.id);
         },
         maxRow() {
             return Math.max(...this.dataWithoutGhostElement.map(item => item.row));
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: I would suggest you to use getMaxObjectValueInArrayByObjectKey from arrayWrapper

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 2]
```diff
@@ -45,202 +45,142 @@ export default {
...
     },
     methods: {
-        calculateRowsCount(time = 300) {
-            debounce(() => {
-                const { clientHeight } = document.querySelector('.grid-container');
-                const visibleRows = Math.ceil(clientHeight / this.rowsHeight);
-                const totalRows = Math.max(this.gridData.length, visibleRows) + 1;
-                this.$emit('setRowsCount', { key: 'rowsCount', value: totalRows });
-            }, time)();
+        calculateRowsCount() {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: Method does too much - would divide emitting event from calculations

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 3]
```diff
@@ -45,202 +45,142 @@ export default {
...
-                        id: this.ghostIndex,
-                        column: this.lastColumn,
-                        row: this.lastRow,
-                        ghost: true,
-                    },
+        onDragOver(event) {
+            const { clientX, clientY } = event;
+            const { overRow, collisionRelation } = this.mousePosition;
+            if (this.onDragFirstItem(event)) return false;
+            const tmpCollisionRelation = this.getMouseOverProps(clientX, clientY);
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: variable is not clear enough

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 4]
```diff
@@ -45,202 +45,142 @@ export default {
...
+            const collidingItem = this.collisionsDetect(overRow);
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: this.collisionsDetect is not clear enough - if the variable is called collidingItem then I would name it as getCollidingItemAtRow

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 5]
```diff
@@ -45,202 +45,142 @@ export default {
...
+            if (collidingItem !== null && collisionRelation !== tmpCollisionRelation) {
+                this.setGhostItemPosition(this.getCollidingPosition(collidingItem));
+            } else if (collidingItem === null && overRow > this.maxRow) {
+                this.setGhostItemPosition({
+                    column: this.setAllowedColumn(),
+                    row: this.maxRow + this.positionBeetwenRows,
                 });
             }
+            event.preventDefault();
+            return true;
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: Return is not required. Why preventing default behaviour ?

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 6]
```diff
@@ -45,202 +45,142 @@ export default {
...
-                        ghost: true,
-                    },
+        onDragOver(event) {
+            const { clientX, clientY } = event;
+            const { overRow, collisionRelation } = this.mousePosition;
+            if (this.onDragFirstItem(event)) return false;
+            const tmpCollisionRelation = this.getMouseOverProps(clientX, clientY);
+            const collidingItem = this.collisionsDetect(overRow);
+
+            if (collidingItem !== null && collisionRelation !== tmpCollisionRelation) {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: conditions are not clear, would use variables to describe them and also I would wrap it into one method

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 7]
```diff
@@ -45,202 +45,142 @@ export default {
...
-                overColumn,
-                overRow,
-                positionOverRow,
-            };
+        onDragLeave(event) {
+            const { fromElement } = event;
+            if (fromElement && fromElement.classList.contains('blur')) {
+                this.removeGhostElement();
+            }
+            event.preventDefault();
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: preventDefault not needed

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 8]
```diff
@@ -45,202 +45,142 @@ export default {
...
-                ? Math.floor(clientY - shadowItem.getBoundingClientRect().top)
-                : null;
-            return {
-                overColumn,
-                overRow,
-                positionOverRow,
-            };
+        onDragLeave(event) {
+            const { fromElement } = event;
+            if (fromElement && fromElement.classList.contains('blur')) {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: it is not going to remove ghost element in each scenario - something might overflow blur - I would rather check the bounds of grid container

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 9]
```diff
@@ -45,202 +45,142 @@ export default {
...
+                column,
+                row,
+                parent: this.setParentId(row, column),
+            });
+            if (withoutGhostLength >= this.rows - 2) {
+                this.$emit('setRowsCount', this.rows + 1);
+            }
+            this.$emit('rebuildGrid', this.draggedElement);
+            event.preventDefault();
+            return true;
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: return true - not required / prevent default cancels event and returning true means that all went ok.

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 10]
```diff
@@ -45,202 +45,142 @@ export default {
...
-            return layerPositionY > centerPosition ? 'son' : 'brother';
+        onDrop(event) {
+            const { row, column } = this.ghostElement;
+            const { length: withoutGhostLength } = this.dataWithoutGhostElement;
+            this.removeGhostElement();
+            this.addItem({
+                id: this.draggedElement,
+                column,
+                row,
+                parent: this.setParentId(row, column),
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: setParentId ? seems like method is returning something - suggest to rename it to get

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 11]
```diff
@@ -45,202 +45,142 @@ export default {
...
+            });
+            if (withoutGhostLength >= this.rows - 2) {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: I would wrapp it into one method - including destructuring

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 12]
```diff
@@ -45,202 +45,142 @@ export default {
...
+            event.preventDefault();
+            return true;
         },
-        isRowSet(row) {
-            const [item] = this.dataWithoutGhostElement.filter(element => element.row === row);
-            return item || null;
+        removeGhostElement() {
+            this.$emit('removeItem', this.ghostElement.id);
+            this.ghostElement.row = null;
+            this.ghostElement.column = null;
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: Strong reference cycle with Vuex store

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 13]
```diff
@@ -45,202 +45,142 @@ export default {
...
+            this.$emit('addItem', item);
         },
-        setAllowedColumn(column, interactionItem, relation = false) {
-            const { column: interactionItemColumn, row: interactionItemRow } = interactionItem;
-            const topBrother = this.isRowSet(interactionItemRow - 1);
-            if (column === null) return null;
-            if (relation) {
-                if (topBrother && column > topBrother.column + 1 && relation === 'brother') return null;
-                return column > interactionItemColumn ? interactionItemColumn + 1 : column;
+        getBottomCollidingColumn({ adjacentElColumn, collidingElColumn }) {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: adjacent - no such word

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 14]
```diff
@@ -45,202 +45,142 @@ export default {
...
-        },
-        setCollapseParent(id) {
-            return Object.keys(this.hiddenItems).reduce((acc, el) => {
-                let x = acc;
-                if (this.hiddenItems[el].some(e => e === id)) {
-                    x = el;
-                }
-                return x;
-            }, '');
+        getTopCollidingColumn({ adjacentElColumn, collidingElColumn }) {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: adjacent - same as above

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 15]
```diff
@@ -45,202 +45,142 @@ export default {
...
-            const item = this.isRowSet(row);
-            if (!item) return null;
-            const hiddenItemParent = this.setCollapseParent(item.id);
-            if (!hiddenItemParent) return null;
-            const collapseItem = this.hiddenItems[hiddenItemParent];
-            const lastCollapseChildId = collapseItem[collapseItem.length - 1];
-            const [lastItem] = this.dataWithoutGhostElement.filter(
-                el => el.id === lastCollapseChildId,
+        getCollidingPosition(collidingEl) {
+            const { collisionRelation } = this.mousePosition;
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: collisionRelation doesnt match to the context of mousePosition object 'mousePosition'

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 16]
```diff
@@ -45,202 +45,142 @@ export default {
...
+            const isTop = collisionRelation === 'top';
+            const [adjacentEl] = this.dataWithoutGhostElement.filter(
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: adjacentEl - no such word

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 17]
```diff
@@ -45,202 +45,142 @@ export default {
...
+                el => el.row === (isTop ? collidingEl.row - 1 : collidingEl.row + 1),
             );
-            return lastItem.row + 0.5 || null;
+            const collidingData = {
+                adjacentElColumn: adjacentEl ? adjacentEl.column : null,
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: adjacent

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 18]
```diff
@@ -45,202 +45,142 @@ export default {
...
+                collidingElColumn: collidingEl.column,
+            };
+            if ((!adjacentEl || collidingEl.row === 0) && isTop) {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: conditions are not readable - assign them to variable

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 19]
```diff
@@ -45,202 +45,142 @@ export default {
...
-            return true;
-        },
-        onDragLeave(event) {
-            // I assume that lastRow and lastColumn has to be cleared when we remove ghost element right ?
-            const { fromElement } = event;
-            if (fromElement && fromElement.classList.contains('blur')) {
-                this.$emit('removeItem', { index: this.lastRow });
-                this.lastRow = null;
-                this.lastColumn = null;
+                e.preventDefault();
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: preventDefault is related to return false

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 20]
```diff
@@ -45,202 +45,142 @@ export default {
...
-            if (!this.ghostElement) {
-                event.stopPropagation();
-                return null;
-            }
-            const { row, column } = this.ghostElement;
-            this.$emit('removeItem', { index: this.lastRow });
-            this.lastRow = null;
-            this.lastColumn = null;
-
+        setParentId(row, column) {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: it is not setting but returning value

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 21]
```diff
@@ -249,42 +189,58 @@ export default {
                 const parent = Math.floor(Math.max(...findElements.map(item => item.row)));
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: getMaxObjectValueInArrayByObjectKey from arrayWrapper

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 22]
```diff
@@ -249,42 +189,58 @@ export default {
...
                 parentId = this.dataWithoutGhostElement[parent].id;
             }
-            if (this.dataWithoutGhostElement.length >= this.rows - 2) {
-                this.$emit('setRowsCount', { key: 'rowsCount', value: this.rows + 1 });
+            return parentId;
+        },
+        setGhostItemPosition({ column, row }) {
+            if (
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: Would assign conditions into variables

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 23]
```diff
@@ -249,42 +189,58 @@ export default {
...
+                (this.ghostElement.column !== column || this.ghostElement.row !== row)
+                && row !== null
+                && column !== null
+            ) {
+                this.ghostElement.row = row;
+                this.ghostElement.column = column;
+                this.addItem({ ...this.ghostElement });
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: Vuex strong reference cycle

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 24]
```diff
@@ -249,42 +189,58 @@ export default {
...
-                            row: e.row,
-                            parent: parentId,
-                        },
-                    });
-                });
+        getMouseOverProps(clientX, clientY) {
+            const elements = document.elementsFromPoint(clientX, clientY);
+            const shadowItem = elements.find(e => e.classList.contains('shadow-grid-item'));
+            const positionOverRow = clientY - shadowItem.getBoundingClientRect().top;
+            const collisionRelation = this.checkCollidingRelation(positionOverRow);
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: collision !== colliding - use one naming convention - collidingRelation

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 25]
```diff
@@ -249,42 +189,58 @@ export default {
...
+
+            if (collisionRelation !== this.mousePosition.collisionRelation) {
+                this.mousePosition.collisionRelation = collisionRelation;
+            }
+            if (shadowItem !== this.mousePosition.shadowItem) {
+                const shadowItemId = shadowItem ? shadowItem.getAttribute('shadow-id') : null;
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: whats the shadow ? Is it ghost element ?

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 26]
```diff
@@ -249,42 +189,58 @@ export default {
...
-                            id: e.id,
-                            column: e.column,
-                            row: e.row,
-                            parent: parentId,
-                        },
-                    });
-                });
+        getMouseOverProps(clientX, clientY) {
+            const elements = document.elementsFromPoint(clientX, clientY);
+            const shadowItem = elements.find(e => e.classList.contains('shadow-grid-item'));
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: shadow item ? ghost element ?

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 27]
```diff
@@ -249,42 +189,58 @@ export default {
...
+            const [item] = this.dataWithoutGhostElement.filter(element => element.row === row);
+            return item || null;
+        },
+        collisionsDetect(row) {
+            const { row: ghostRow } = this.ghostElement;
+            if (ghostRow !== null && row === ghostRow + this.positionBeetwenRows) return null;
+            const newRow = ghostRow !== null && row > ghostRow
+                ? row - 1
+                : row;
+            return this.isRowSet(newRow);
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: isRowSet seems like computed property not a getter

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 28]
```diff
@@ -249,42 +189,58 @@ export default {
...
+                    shadowItem,
+                };
             }
+            return collisionRelation;
+        },
+        checkCollidingRelation(layerPositionY) {
+            const centerPosition = Math.floor(this.rowsHeight / 2);
+            return layerPositionY > centerPosition ? 'bottom' : 'top';
+        },
+        isRowSet(row) {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: Seems like getter

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 29]
```diff
@@ -249,42 +189,58 @@ export default {
...
+            const [item] = this.dataWithoutGhostElement.filter(element => element.row === row);
+            return item || null;
+        },
+        collisionsDetect(row) {
+            const { row: ghostRow } = this.ghostElement;
+            if (ghostRow !== null && row === ghostRow + this.positionBeetwenRows) return null;
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: Describe condition with variable

File: components/TemplateGrid/TemplateGridContainer.vue [hunk 30]
```diff
@@ -249,42 +189,58 @@ export default {
...
+            const newRow = ghostRow !== null && row > ghostRow
+                ? row - 1
+                : row;
+            return this.isRowSet(newRow);
+        },
+        setAllowedColumn() {
```
Comment (derpdead) [inline components/TemplateGrid/TemplateGridContainer.vue]: Getter

File: store/tree/actions.js [hunk 31]
```diff
@@ -5,42 +5,36 @@
...
-        commit('removeTreeItem', { index });
+    removeTreeItem: ({ commit }, id) => {
+        commit('removeTreeItem', id);
     },
-    rebuildTree: ({ commit, state }, { id }) => {
+    rebuildTree: ({ commit, state }, id) => {
         const findIndex = state.treeData.findIndex(el => el.id === id);
         const newTree = state.treeData.reduce((accumulator, current, i) => {
-            if (i === findIndex) {
+            if (i === findIndex && current.row !== 0) {
```
Comment (derpdead) [inline store/tree/actions.js]: Describe condition with variable

File: components/TemplateGrid/TemplateGridContainer.vue (hunk 26 above)
Comment (bleto) [reply components/TemplateGrid/TemplateGridContainer.vue]: Shadow-item is the element responsible for displaying the logic of grid presentation and has nothing to do with ghost-eleemnt

File: store/tree/actions.js (hunk 31 above)
Comment (tomaszstrozik) [reply store/tree/actions.js]: I think you can leave it, it's short enough to understand what is going on

File: components/TemplateGrid/TemplateGridContainer.vue (hunk 11 above)
Comment (bleto) [reply components/TemplateGrid/TemplateGridContainer.vue]: I think it not necessary

File: components/TemplateGrid/TemplateGridContainer.vue (hunk 25 above)
Comment (bleto) [reply components/TemplateGrid/TemplateGridContainer.vue]: The answer is higher.

**Pull request merged**